.github/actions/github-check-upcoming-runs/action.yml

```diff
@@ -0,0 +1,54 @@
+name: Fail if an upcoming run is detected
+description: Exits on error when an upcoming run is detected on the same branch, and rerun it if cancelled
+inputs:
+  github-token:
+    description: GitHub Token
+    required: true
+  workflow:
+    description: The workflow file name eg workflow.yml
+    required: true
+
+outputs:
+  exit:
+    description: True if this check produced an exit
+    value: ${{ steps.check.outputs.exit }}
+
+runs:
+  using: "composite"
+  steps:
+    - name: Get branch name
+      uses: Alfresco/alfresco-build-tools/.github/actions/get-branch-name@469ed8f7c8198c4f8476392d944b2568e4043c5a # v5.13.1
+
+    - name: Check upcoming runs
+      id: check
+      env:
+        GH_TOKEN: ${{ inputs.github-token }}
+        WORKFLOW: ${{ inputs.workflow }}
+      shell: bash
+      run: |
+        URL="/repos/$GITHUB_REPOSITORY/actions/workflows/$WORKFLOW/runs"
+        echo "Checking runs at $URL for $BRANCH_NAME"
+        RUNS=$(gh api -X GET $URL -F branch=$BRANCH_NAME -F event=pull_request)
+
+        RUN=$(echo $RUNS | jq --argjson id $GITHUB_RUN_ID 'first(.workflow_runs[] | select(.id > $id) | {"id":.id, "status":.status, "conclusion":.conclusion})')
+        if [ ! -z "$RUN" ]; then
+
+          echo "Current run id: $GITHUB_RUN_ID"
+          echo "Upcoming runs:"
+          echo $RUNS | jq --argjson id $GITHUB_RUN_ID '.workflow_runs[] | select(.id > $id) | {"id":.id, "status":.status, "conclusion":.conclusion}'
+
+          ID=$(echo $RUN | jq -r ".id")
+          STATUS=$(echo $RUN | jq -r ".status")
+          CONCLUSION=$(echo $RUN | jq -r ".conclusion")
+          RERUN_URL="$GITHUB_SERVER_URL/$GITHUB_REPOSITORY/actions/runs/$ID"
+
+          if [[ "$STATUS" == "completed" && "$CONCLUSION" == "cancelled" ]]; then
+            echo "::notice::Re-running $ID: $RERUN_URL"
+            gh run rerun -R $GITHUB_REPOSITORY $ID --failed
```

Comment:
`--failed` is really needed? if the workflow has concurrency and is found cancelled, probably no job has ever run?

Reply:
do you think it would fail if so? it's just to shortcut re-running every job that has gone through, but I can make this optional if you prefer, I do not expect it to fail if it was fully cancelled